### PR TITLE
Add Merge Sort Clear Version

### PR DESCRIPTION
**Pull request by 0xsuu**

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/Llama-Group/Project-Llama/pull/145" target="_blank">ReviewNinja

File: llama_cpp/src/Sort/MergeSort.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+namespace llama {
+template <class T> class MergeSort : public Sort<T> {
+ public:
+    MergeSort() {}
+    explicit MergeSort(std::vector<T> *inputArray) {
+    this->performSort(inputArray);
+    }
+
+    void Merge(std::vector<SortObject<T>> *arr, int low, int mid, int high) {
+      int i = low;
```
Comment (0xsuu) [inline llama_cpp/src/Sort/MergeSort.h]: space

File: llama_cpp/src/Sort/MergeSort.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    void mergesort(std::vector<SortObject<T>> *arr, int low, int high) {
+      int mid;
+      if (low < high) {
+      mid = (low + high) / 2;
+      // Left hand side。
+      mergesort(arr, low, mid);
+      // Right hand side.
+      mergesort(arr, mid + 1, high);
+      Merge(arr, low, mid, high);
+      }
```
Comment (0xsuu) [inline llama_cpp/src/Sort/MergeSort.h]: SPACE!! This is not acceptable!

File: llama_cpp/src/Sort/MergeSort.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+#include <vector>
+
+#include "Sort.h"
+
+namespace llama {
+template <class T> class MergeSort : public Sort<T> {
+ public:
+    MergeSort() {}
+    explicit MergeSort(std::vector<T> *inputArray) {
+    this->performSort(inputArray);
```
Comment (BlueCocoa) [inline llama_cpp/src/Sort/MergeSort.h]: space

Comment (0xsuu) [reply llama_cpp/src/Sort/MergeSort.h]: 233

File: llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp [hunk 4]
```diff
@@ -309,7 +314,7 @@ TEST_F(SortCorrectnessTest, BubbleRandomStringsCorrectness) {
     EXPECT_EQ(randomStringsCorrect, randomStrings);
 }
 
-// Testing for Selection Sort.
+// Testing for Selection Sort.
```
Comment (0xsuu) [inline llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp]: Remove white space at the end.

File: llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp [hunk 5]
```diff
@@ -293,7 +298,7 @@ TEST_F(SortCorrectnessTest, InsertionRandomStringsCorrectness) {
     EXPECT_EQ(randomStringsCorrect, randomStrings);
 }
 
-// Testing for Bubble Sort.
+// Testing for Bubble Sort.
```
Comment (0xsuu) [inline llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp]: Remove white space at the end.

File: llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp [hunk 6]
```diff
@@ -277,7 +282,7 @@ TEST_F(SortCorrectnessTest, BruteForceRandomStringsCorrectness) {
     EXPECT_EQ(randomStringsCorrect, randomStrings);
 }
 
-// Testing for Insertion Sort.
+// Testing for Insertion Sort.
```
Comment (0xsuu) [inline llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp]: Remove white space at the end.

File: llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp [hunk 7]
```diff
@@ -261,7 +266,7 @@ struct SortCorrectnessTest : public ::testing::Test {
     }
 };
 
-// Testing for Brute Force Sort.
+// Testing for Brute Force Sort.
```
Comment (0xsuu) [inline llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp]: Remove white space at the end.

File: llama_cpp/src/Sort/MergeSort.h [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+#include <algorithm>
+#include <utility>
+#include <vector>
+
+#include "Sort.h"
+
+namespace llama {
+template <class T> class MergeSort : public Sort<T> {
+ public:
+        MergeSort() {}
```
Comment (0xsuu) [inline llama_cpp/src/Sort/MergeSort.h]: 4 spaces here not 8

File: llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp (hunk 7 above)
Comment (0xsuu) [reply llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp]: Still not fixed.

File: llama_cpp/src/Sort/MergeSort.h [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+            std::swap(tmp.at(k++), arr->at(i++));
+        }
+        while (j <= high) {
+            std::swap(tmp.at(k++), arr->at(j++));
+        }
+        for (i = low, k = 0; i <= high; i++, k++) {
+            std::swap(tmp.at(k), arr->at(i));
+        }
+        std::vector<SortObject<T>>().swap(tmp);
+        }
```
Comment (0xsuu) [inline llama_cpp/src/Sort/MergeSort.h]: This bracket is not in right place.

**Comment by 0xsuu** [comment]
LGTM

**Comment by BlueCocoa** [comment]
LGTM

**Pull request merged**